File: stack/leetcode/946.cpp

// stack and two pointer (pushed, popped)
int stack[1002];
int top;
bool validateStackSequences(int* pushed, int pushedSize, int* popped, int poppedSize){
    top = 0;
    int j = 0; // index of popped
    for (int  i = 0; i < pushedSize; i++){
        stack[top++] = pushed[i];
        while (top > 0 && stack[top - 1] == popped[j]){
            top--;
            j++;
        }
    }
    return j == poppedSize;
}

// https://leetcode.com/problems/validate-stack-sequences/

File: stack/leetcode/946/946.py
def validateStackSequences(pushed: list[int], popped: list[int]) -> bool:
    stack = []
    index = 0
    for num in pushed:
        stack.append(num)
        while (stack and stack[-1] == popped[index]):
            stack.pop()
            index += 1
    if (stack):
        return False
    else:
        return True